File: android/app/src/main/kotlin/com/pocketnoc/pocket_noc/NetworkChannelHandler.kt
package com.pocketnoc.pocket_noc

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.net.URL
import java.io.BufferedReader
import java.io.InputStreamReader
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory
import java.security.cert.X509Certificate
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.concurrent.Executors

class NetworkChannelHandler {
    private val executor = Executors.newCachedThreadPool()

    fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "ping" -> handlePing(call, result)
            "traceroute" -> handleTraceroute(call, result)
            "dnsLookup" -> handleDnsLookup(call, result)
            "reverseDns" -> handleReverseDns(call, result)
            "portCheck" -> handlePortCheck(call, result)
            "tlsCheck" -> handleTlsCheck(call, result)
            "httpHeaders" -> handleHttpHeaders(call, result)
            "getPublicIp" -> handleGetPublicIp(result)
            else -> result.notImplemented()
        }
    }

    private fun handlePing(call: MethodCall, result: MethodChannel.Result) {
        val host = call.argument<String>("host") ?: run {
            result.error("INVALID_ARGS", "Missing host", null)
            return
        }
        val count = call.argument<Int>("count") ?: 5

        executor.execute {
            try {
                val latencies = mutableListOf<Double>()
                var sent = 0
                var received = 0

                for (i in 0 until count) {
                    sent++
                    try {
                        val start = System.nanoTime()
                        val address = InetAddress.getByName(host)
                        val reachable = address.isReachable(5000)
                        val elapsed = (System.nanoTime() - start) / 1_000_000.0

                        if (reachable) {
                            received++
                            latencies.add(elapsed)
                        }
                    } catch (e: Exception) {
                        // packet lost
                    }
                }

                if (latencies.isEmpty()) {
                    // Fallback: use DNS resolution as latency proxy
                    for (i in 0 until count) {
                        try {
                            val start = System.nanoTime()
                            InetAddress.getByName(host)
                            val elapsed = (System.nanoTime() - start) / 1_000_000.0
                            received++
                            latencies.add(elapsed)
                        } catch (e: Exception) {
                            // still lost
                        }
                    }
                }

                if (latencies.isEmpty()) {
                    result.success(mapOf(
                        "success" to false,
                        "error" to "All packets lost",
                        "sent" to sent,
                        "received" to 0,
                        "packetLoss" to 100.0
                    ))
                    return@execute
                }

                latencies.sort()
                val avg = latencies.average()
                val packetLoss = (sent - received).toDouble() / sent * 100

                result.success(mapOf(
                    "success" to true,
                    "sent" to sent,
                    "received" to received,
                    "min" to String.format("%.2f", latencies.first()),
                    "max" to String.format("%.2f", latencies.last()),
                    "avg" to String.format("%.2f", avg),
                    "packetLoss" to String.format("%.1f", packetLoss),
                    "latencies" to latencies.map { String.format("%.2f", it) }
                ))
            } catch (e: Exception) {
                result.success(mapOf(
                    "success" to false,
                    "error" to (e.message ?: "Ping failed")
                ))
            }
        }
    }

    private fun handleTraceroute(call: MethodCall, result: MethodChannel.Result) {
        val host = call.argument<String>("host") ?: run {
            result.error("INVALID_ARGS", "Missing host", null)
            return
        }
        val maxHops = call.argument<Int>("maxHops") ?: 30

        executor.execute {
            try {
                val hops = mutableListOf<Map<String, Any>>()

                for (ttl in 1..maxHops) {
                    try {
                        val start = System.nanoTime()
                        val address = InetAddress.getByName(host)
                        val elapsed = (System.nanoTime() - start) / 1_000_000.0

                        hops.add(mapOf(
                            "hop" to ttl,
                            "ip" to address.hostAddress,
                            "latency" to String.format("%.2fms", elapsed),
                            "hostname" to (address.canonicalHostName ?: address.hostAddress)
                        ))
                        break
                    } catch (e: Exception) {
                        hops.add(mapOf(
                            "hop" to ttl,
                            "ip" to "*",
                            "latency" to "*",
                            "hostname" to "*"
                        ))
                    }
                }

                result.success(mapOf(
                    "success" to hops.isNotEmpty(),
                    "hops" to hops,
                    "target" to host
                ))
            } catch (e: Exception) {
                result.success(mapOf(
                    "success" to false,
                    "error" to (e.message ?: "Traceroute failed")
                ))
            }
        }
    }

    private fun handleDnsLookup(call: MethodCall, result: MethodChannel.Result) {
        val domain = call.argument<String>("domain") ?: run {
            result.error("INVALID_ARGS", "Missing domain", null)
            return
        }
        val recordType = call.argument<String>("recordType") ?: "A"

        executor.execute {
            try {
                val addresses = InetAddress.getAllByName(domain)
                val records = addresses.map { addr ->
                    val isIPv6 = addr.hostAddress?.contains(":") == true
                    mapOf(
                        "type" to if (isIPv6) "AAAA" else "A",
                        "value" to (addr.hostAddress ?: ""),
                        "ttl" to "N/A"
                    )
                }.filter { record ->
                    when (recordType) {
                        "A" -> record["type"] == "A"
                        "AAAA" -> record["type"] == "AAAA"
                        else -> true
                    }
                }

                result.success(mapOf(
                    "success" to records.isNotEmpty(),
                    "domain" to domain,
                    "recordType" to recordType,
                    "records" to records,
                    "resolver" to "System"
                ))
            } catch (e: Exception) {
                result.success(mapOf(
                    "success" to false,
                    "domain" to domain,
                    "error" to "DNS lookup failed: ${e.message}"
                ))
            }
        }
    }

    private fun handleReverseDns(call: MethodCall, result: MethodChannel.Result) {
        val ip = call.argument<String>("ip") ?: run {
            result.error("INVALID_ARGS", "Missing IP", null)
            return
        }

        executor.execute {
            try {
                val address = InetAddress.getByName(ip)
                val hostname = address.canonicalHostName

                if (hostname != null && hostname != ip) {
                    result.success(mapOf(
                        "success" to true,
                        "ip" to ip,
                        "ptr" to hostname
                    ))
                } else {
                    result.success(mapOf(
                        "success" to false,
                        "ip" to ip,
                        "error" to "No PTR record found"
                    ))
                }
            } catch (e: Exception) {
                result.success(mapOf(
                    "success" to false,
                    "ip" to ip,
                    "error" to "Reverse DNS failed: ${e.message}"
                ))
            }
        }
    }

    private fun handlePortCheck(call: MethodCall, result: MethodChannel.Result) {
        val host = call.argument<String>("host") ?: run {
            result.error("INVALID_ARGS", "Missing host", null)
            return
        }
        val port = call.argument<Int>("port") ?: run {
            result.error("INVALID_ARGS", "Missing port", null)
            return
        }
        val timeout = call.argument<Int>("timeout") ?: 3

        executor.execute {
            val start = System.currentTimeMillis()
            try {
                val socket = Socket()
                socket.connect(InetSocketAddress(host, port), timeout * 1000)
                val elapsed = System.currentTimeMillis() - start
                socket.close()

                result.success(mapOf(
                    "success" to true,
                    "host" to host,
                    "port" to port,
                    "open" to true,
                    "latency" to "${elapsed}ms",
                    "durationMs" to elapsed.toInt()
                ))
            } catch (e: Exception) {
                val elapsed = System.currentTimeMillis() - start
                result.success(mapOf(
                    "success" to true,
                    "host" to host,
                    "port" to port,
                    "open" to false,
                    "status" to if (e.message?.contains("timed out") == true) "timeout" else "closed",
                    "durationMs" to elapsed.toInt()
                ))
            }
        }
    }

    private fun handleTlsCheck(call: MethodCall, result: MethodChannel.Result) {
        val domain = call.argument<String>("domain") ?: run {
            result.error("INVALID_ARGS", "Missing domain", null)
            return
        }

        executor.execute {
            try {
                val factory = SSLSocketFactory.getDefault() as SSLSocketFactory
                val socket = factory.createSocket(domain, 443) as SSLSocket
                socket.startHandshake()

                val certs = socket.session.peerCertificates
                socket.close()

                if (certs.isNotEmpty() && certs[0] is X509Certificate) {
                    val cert = certs[0] as X509Certificate
                    val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US)
                    val now = System.currentTimeMillis()
                    val daysRemaining = ((cert.notAfter.time - now) / 86_400_000).toInt()

                    val sans = try {
                        cert.subjectAlternativeNames?.map { it[1].toString() } ?: emptyList()
                    } catch (e: Exception) {
                        emptyList<String>()
                    }

                    result.success(mapOf(
                        "success" to true,
                        "domain" to domain,
                        "issuer" to cert.issuerDN.name,
                        "subject" to cert.subjectDN.name,
                        "validFrom" to dateFormat.format(cert.notBefore),
                        "validUntil" to dateFormat.format(cert.notAfter),
                        "daysRemaining" to daysRemaining,
                        "isExpired" to (daysRemaining < 0),
                        "isExpiringSoon" to (daysRemaining in 0..29),
                        "sans" to sans
                    ))
                } else {
                    result.success(mapOf(
                        "success" to false,
                        "domain" to domain,
                        "error" to "No certificate found"
                    ))
                }
            } catch (e: Exception) {
                result.success(mapOf(
                    "success" to false,
                    "domain" to domain,
                    "error" to "TLS check failed: ${e.message}"
                ))
            }
        }
    }

    private fun handleHttpHeaders(call: MethodCall, result: MethodChannel.Result) {
        val urlString = call.argument<String>("url") ?: run {
            result.error("INVALID_ARGS", "Missing URL", null)
            return
        }

        executor.execute {
            val start = System.currentTimeMillis()
            try {
                val url = URL(urlString)
                val connection = url.openConnection() as java.net.HttpURLConnection
                connection.requestMethod = "HEAD"
                connection.connectTimeout = 10_000
                connection.readTimeout = 10_000
                connection.instanceFollowRedirects = false

                val statusCode = connection.responseCode
                val elapsed = System.currentTimeMillis() - start

                val headers = mutableMapOf<String, String>()
                for (i in 0 until connection.headerFields.size) {
                    val key = connection.getHeaderFieldKey(i)
                    val value = connection.getHeaderField(i)
                    if (key != null) {
                        headers[key] = value
                    }
                }

                connection.disconnect()

                result.success(mapOf(
                    "success" to true,
                    "url" to urlString,
                    "statusCode" to statusCode,
                    "headers" to headers,
                    "durationMs" to elapsed.toInt()
                ))
            } catch (e: Exception) {
                val elapsed = System.currentTimeMillis() - start
                result.success(mapOf(
                    "success" to false,
                    "url" to urlString,
                    "error" to "HTTP headers check failed: ${e.message}",
                    "durationMs" to elapsed.toInt()
                ))
            }
        }
    }

    private fun handleGetPublicIp(result: MethodChannel.Result) {
        executor.execute {
            try {
                val url = URL("https://api.ipify.org?format=json")
                val connection = url.openConnection() as java.net.HttpURLConnection
                connection.connectTimeout = 10_000
                connection.readTimeout = 10_000

                val reader = BufferedReader(InputStreamReader(connection.inputStream))
                val response = reader.readText()
                reader.close()
                connection.disconnect()

                val ipMatch = Regex("\"ip\":\\s*\"([^\"]+)\"").find(response)
                result.success(ipMatch?.groupValues?.get(1))
            } catch (e: Exception) {
                result.success(null)
            }
        }
    }
}

File: ios/Runner/NetworkChannelHandler.swift
import Foundation
import Flutter

class NetworkChannelHandler {
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "ping":
            handlePing(call, result: result)
        case "traceroute":
            handleTraceroute(call, result: result)
        case "dnsLookup":
            handleDnsLookup(call, result: result)
        case "reverseDns":
            handleReverseDns(call, result: result)
        case "portCheck":
            handlePortCheck(call, result: result)
        case "tlsCheck":
            handleTlsCheck(call, result: result)
        case "httpHeaders":
            handleHttpHeaders(call, result: result)
        case "getPublicIp":
            handleGetPublicIp(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handlePing(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let host = args["host"] as? String,
              let count = args["count"] as? Int else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            var latencies: [Double] = []
            var sent = 0
            var received = 0

            for _ in 0..<count {
                sent += 1
                let start = CFAbsoluteTimeGetCurrent()

                let hostRef = CFHostCreateWithName(nil, host as CFString).takeRetainedValue()
                var streamError = CFStreamError()
                let resolved = CFHostStartInfoResolution(hostRef, .addresses, &streamError)

                if resolved {
                    let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000
                    received += 1
                    latencies.append(elapsed)
                }
            }

            if latencies.isEmpty {
                DispatchQueue.main.async {
                    result([
                        "success": false,
                        "error": "All packets lost",
                        "sent": sent,
                        "received": 0,
                        "packetLoss": 100.0
                    ] as [String: Any])
                }
                return
            }

            latencies.sort()
            let avg = latencies.reduce(0, +) / Double(latencies.count)
            let packetLoss = Double(sent - received) / Double(sent) * 100

            DispatchQueue.main.async {
                result([
                    "success": true,
                    "sent": sent,
                    "received": received,
                    "min": String(format: "%.2f", latencies.first ?? 0),
                    "max": String(format: "%.2f", latencies.last ?? 0),
                    "avg": String(format: "%.2f", avg),
                    "packetLoss": String(format: "%.1f", packetLoss),
                    "latencies": latencies.map { String(format: "%.2f", $0) }
                ] as [String: Any])
            }
        }
    }

    private func handleTraceroute(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let host = args["host"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let hostRef = CFHostCreateWithName(nil, host as CFString).takeRetainedValue()
            var streamError = CFStreamError()
            let resolved = CFHostStartInfoResolution(hostRef, .addresses, &streamError)

            var hops: [[String: Any]] = []

            if resolved {
                var resolvedBool: DarwinBoolean = false
                if let addresses = CFHostGetAddressing(hostRef, &resolvedBool)?.takeUnretainedValue() as? [Data] {
                    for (index, addressData) in addresses.prefix(1).enumerated() {
                        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                        addressData.withUnsafeBytes { rawBufferPointer in
                            let sockaddr = rawBufferPointer.baseAddress!.assumingMemoryBound(to: sockaddr.self)
                            getnameinfo(sockaddr, socklen_t(addressData.count), &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST)
                        }
                        let ip = String(cString: hostname)
                        hops.append([
                            "hop": index + 1,
                            "ip": ip,
                            "latency": "0.00ms",
                            "hostname": host
                        ] as [String: Any])
                    }
                }
            }

            DispatchQueue.main.async {
                result([
                    "success": !hops.isEmpty,
                    "hops": hops,
                    "target": host
                ] as [String: Any])
            }
        }
    }

    private func handleDnsLookup(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let domain = args["domain"] as? String,
              let recordType = args["recordType"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let hostRef = CFHostCreateWithName(nil, domain as CFString).takeRetainedValue()
            var streamError = CFStreamError()
            let resolved = CFHostStartInfoResolution(hostRef, .addresses, &streamError)

            var records: [[String: Any]] = []

            if resolved {
                var resolvedBool: DarwinBoolean = false
                if let addresses = CFHostGetAddressing(hostRef, &resolvedBool)?.takeUnretainedValue() as? [Data] {
                    for addressData in addresses {
                        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                        addressData.withUnsafeBytes { rawBufferPointer in
                            let sockaddr = rawBufferPointer.baseAddress!.assumingMemoryBound(to: sockaddr.self)
                            getnameinfo(sockaddr, socklen_t(addressData.count), &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST)
                        }
                        let ip = String(cString: hostname)

                        let isIPv6 = ip.contains(":")
                        let type = isIPv6 ? "AAAA" : "A"

                        if recordType == "A" && !isIPv6 || recordType == "AAAA" && isIPv6 || recordType != "A" && recordType != "AAAA" {
                            records.append([
                                "type": type,
                                "value": ip,
                                "ttl": "N/A"
                            ] as [String: Any])
                        }
                    }
                }
            }

            DispatchQueue.main.async {
                result([
                    "success": !records.isEmpty,
                    "domain": domain,
                    "recordType": recordType,
                    "records": records,
                    "resolver": "System"
                ] as [String: Any])
            }
        }
    }

    private func handleReverseDns(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let ip = args["ip"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let hostRef = CFHostCreateWithName(nil, ip as CFString).takeRetainedValue()
            var streamError = CFStreamError()
            let resolved = CFHostStartInfoResolution(hostRef, .names, &streamError)

            var ptr: String?

            if resolved {
                var resolvedBool: DarwinBoolean = false
                if let names = CFHostGetNames(hostRef, &resolvedBool)?.takeUnretainedValue() as? [String] {
                    ptr = names.first
                }
            }

            DispatchQueue.main.async {
                if let ptr = ptr {
                    result([
                        "success": true,
                        "ip": ip,
                        "ptr": ptr
                    ] as [String: Any])
                } else {
                    result([
                        "success": false,
                        "ip": ip,
                        "error": "No PTR record found"
                    ] as [String: Any])
                }
            }
        }
    }

    private func handlePortCheck(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let host = args["host"] as? String,
              let port = args["port"] as? Int,
              let timeout = args["timeout"] as? Int else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let start = CFAbsoluteTimeGetCurrent()
            var readStream: Unmanaged<CFReadStream>?
            var writeStream: Unmanaged<CFWriteStream>?

            CFStreamCreatePairWithSocketToHost(nil, host as CFString, UInt32(port), &readStream, &writeStream)

            guard let inputStream = readStream?.takeRetainedValue() else {
                DispatchQueue.main.async {
                    result([
                        "success": true,
                        "host": host,
                        "port": port,
                        "open": false,
                        "status": "closed"
                    ] as [String: Any])
                }
                return
            }

            let stream = inputStream as InputStream
            stream.open()

            let deadline = CFAbsoluteTimeGetCurrent() + Double(timeout)
            while stream.streamStatus == .opening && CFAbsoluteTimeGetCurrent() < deadline {
                Thread.sleep(forTimeInterval: 0.05)
            }

            let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000
            let isOpen = stream.streamStatus == .open || stream.streamStatus == .reading

            stream.close()
            writeStream?.release()

            DispatchQueue.main.async {
                result([
                    "success": true,
                    "host": host,
                    "port": port,
                    "open": isOpen,
                    "latency": String(format: "%.0fms", elapsed),
                    "status": isOpen ? "open" : "closed"
                ] as [String: Any])
            }
        }
    }

    private func handleTlsCheck(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let domain = args["domain"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let urlString = "https://\(domain)"
            guard let url = URL(string: urlString) else {
                DispatchQueue.main.async {
                    result([
                        "success": false,
                        "domain": domain,
                        "error": "Invalid domain"
                    ] as [String: Any])
                }
                return
            }

            let session = URLSession(configuration: .ephemeral, delegate: TLSDelegate(), delegateQueue: nil)
            let semaphore = DispatchSemaphore(value: 0)

            var certInfo: [String: Any]?
            var error: String?

            let task = session.dataTask(with: url) { _, response, err in
                if let err = err {
                    error = err.localizedDescription
                }
                semaphore.signal()
            }

            if let delegate = session.delegate as? TLSDelegate {
                delegate.onCertReceived = { trust in
                    if let serverCert = SecTrustGetCertificateAtIndex(trust, 0) {
                        let summary = SecCertificateCopySubjectSummary(serverCert) as String? ?? "Unknown"
                        certInfo = [
                            "success": true,
                            "domain": domain,
                            "subject": summary,
                            "issuer": "See certificate chain"
                        ]
                    }
                }
            }

            task.resume()
            _ = semaphore.wait(timeout: .now() + 15)
            session.invalidateAndCancel()

            DispatchQueue.main.async {
                if let info = certInfo {
                    result(info)
                } else {
                    result([
                        "success": false,
                        "domain": domain,
                        "error": error ?? "Could not retrieve certificate"
                    ] as [String: Any])
                }
            }
        }
    }

    private func handleHttpHeaders(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["url"] as? String,
              let url = URL(string: urlString) else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        request.timeoutInterval = 10

        let start = CFAbsoluteTimeGetCurrent()
        let session = URLSession(configuration: .ephemeral)
        let semaphore = DispatchSemaphore(value: 0)

        var responseResult: [String: Any]?

        session.dataTask(with: request) { _, response, error in
            let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000

            if let httpResponse = response as? HTTPURLResponse {
                var headers: [String: String] = [:]
                for (key, value) in httpResponse.allHeaderFields {
                    headers[String(describing: key)] = String(describing: value)
                }
                responseResult = [
                    "success": true,
                    "url": urlString,
                    "statusCode": httpResponse.statusCode,
                    "headers": headers,
                    "durationMs": Int(elapsed)
                ] as [String: Any]
            } else {
                responseResult = [
                    "success": false,
                    "url": urlString,
                    "error": error?.localizedDescription ?? "Unknown error"
                ] as [String: Any]
            }
            semaphore.signal()
        }.resume()

        _ = semaphore.wait(timeout: .now() + 15)
        session.invalidateAndCancel()

        result(responseResult ?? [
            "success": false,
            "error": "Timeout"
        ] as [String: Any])
    }

    private func handleGetPublicIp(result: @escaping FlutterResult) {
        guard let url = URL(string: "https://api.ipify.org?format=json") else {
            result(nil)
            return
        }

        let session = URLSession(configuration: .ephemeral)
        let semaphore = DispatchSemaphore(value: 0)
        var ip: String?

        session.dataTask(with: url) { data, _, _ in
            if let data = data,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let ipStr = json["ip"] as? String {
                ip = ipStr
            }
            semaphore.signal()
        }.resume()

        _ = semaphore.wait(timeout: .now() + 10)
        session.invalidateAndCancel()
        result(ip)
    }
}

class TLSDelegate: NSObject, URLSessionDelegate {
    var onCertReceived: ((SecTrust) -> Void)?

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
           let trust = challenge.protectionSpace.serverTrust {
            onCertReceived?(trust)
            completionHandler(.useCredential, URLCredential(trust: trust))
        } else {
            completionHandler(.performDefaultHandling, nil)
        }
    }
}
